runtime/src/parachains.rs

```diff
@@ -471,27 +437,29 @@ impl<T: Trait> Module<T> {
 			});
 			// Should never be able to fail assuming our state is uncorrupted, but best not
 			// to panic, even if it does.
-			let _ = <RelayDispatchQueue>::append(id, upward_messages);
+			let _ = RelayDispatchQueue::append(id, upward_messages);
+			if ordered_needs_dispatch.binary_search(&id).is_err() {
+				// same.
+				let _ = NeedsDispatch::append(&[id]);
```

Comment:
if the invariant that `NeedsDispatch` should be sorted is to be maintained, you should `let Err(idx) = binary_search(...)` and insert at the index, rather than appending.